src/store/store_compressor.rs

```diff
@@ -0,0 +1,267 @@
+use std::io::Write;
+use std::sync::mpsc::{sync_channel, Receiver, SyncSender};
+use std::thread::JoinHandle;
+use std::{io, thread};
+
+use common::{BinarySerializable, CountingWriter, TerminatingWrite};
+
+use crate::directory::WritePtr;
+use crate::store::footer::DocStoreFooter;
+use crate::store::index::{Checkpoint, SkipIndexBuilder};
+use crate::store::{Compressor, Decompressor, StoreReader};
+use crate::DocId;
+
+pub struct BlockCompressor(Impls);
+
+// The struct wrapping an enum is just here to keep the
+// impls private.
+enum Impls {
+    SameThread(BlockCompressorImpl),
+    DedicatedThread(DedicatedThreadBlockCompressorImpl),
```

Comment:
I think the naming could be improved here a little
```suggestion
enum BlockCompressorVariants {
    SameThread(SameThreadBlockCompressor),
    DedicatedThread(DedicatedThreadBlockCompressor),
```